src/OpenTelemetry/Trace/EnrichmentScope.cs

```diff
@@ -0,0 +1,78 @@
+﻿// <copyright file="EnrichmentScope.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using OpenTelemetry.Context;
+
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// A class for enriching the data of <see cref="Activity"/> objects.
+    /// </summary>
+    public class EnrichmentScope : IDisposable
```

Comment:
I know I'm always bad at naming things, the name `ContextualProcessor` keeps popping up in my mind 😂

Reply:
(I'm not opposed to changing the name. I was just waiting to see if anyone else thought similar or had other ideas.)

Reply:
Should be my last question in this PR: do we envision a similar concept would be introduced for metrics/logs in the future? If yes, we might need to consider how we name things so we won't be cornered in the future.

Reply:
This is an interesting point. Slight tangent, but my mental model of "instrumentation" before OTel has been one that combines the concepts of collecting of traces, metrics, etc together. High-level I think of something like HTTP instrumentation as a thing that both generates spans in a trace as well as collecting metrics about the operation. From a configuration perspective, I've wondered if these things will continue to be treated as separate - that is, I build up my trace pipeline/configuration separate from my metric configuration.

In the context of EnrichmentScope, should there be a different trace vs. metric vs. log EnrichmentScope concept? Or can all of these be combined in one EnrichmentScope class that includes the ability to configure behavior on traces, metrics, and logs?

Reply:
There's nothing really Activity-specific about the current implementation. It uses @reyang's `RuntimeContext` stuff. Essentially what it is doing is buffering a callback/closure to be executed when the final _thing_ to be enriched is available. So in that way, it could work for metrics & logs. Maybe you will be able to give it different callbacks depending on what is being enriched? I don't have enough of a mental picture of the metrics & logging APIs to say conclusively, but it would be great if we had one thing for all the spec areas. Open to suggestions.

Reply:
My suggestion is that we get this PR merged and continue the discussion here 😆

Reply:
Yes, I'm leaning towards thinking one thing for all the spec areas make sense, though like you my mental picture is still somewhat fuzzy.

This is why I was pondering this thought in the realm of configuration since that's a little more concrete at the moment. For example, I can see the `AddOpenTelemetry` extension method on `IServiceCollection` include additional parameters for a "MeterProviderBuilder" and a "LogProviderBuilder" like how you're suggesting distinct callbacks based on type of telemetry data for EnrichmentScope.

Reply:
> My suggestion is that we get this PR merged and continue the discussion here 😆

yes! agreed.